lib/_stream_readable.js

```diff
@@ -1215,17 +1215,33 @@ function endReadableNT(state, stream) {
     state.endEmitted = true;
     stream.emit('end');
 
+    if (stream.writable && stream.allowHalfOpen === false) {
+      process.nextTick(endWritableNT, state, stream);
+    }
+
     if (state.autoDestroy) {
       // In case of duplex streams we need a way to detect
       // if the writable side is ready for autoDestroy as well
       const wState = stream._writableState;
-      if (!wState || (wState.autoDestroy && wState.finished)) {
+      const autoDestroy = !wState || (
+        wState.autoDestroy &&
+        !stream.writable &&
+        !wState.length
+      );
```

Comment:
This aligns Duplex behavior with the `net.Socket` behavior.

Reply:
Can you add some tests for this? What is the exact change here?

Reply:
It will assume that the stream can be destroyed if `writable` has been explicitily set to `false`. This allows implementations such as `net.Socket` to flag that the stream will never become writable without having to call `end()` and causing side effects such as `'finish'`.

Reply:
Added test